In [33]:
import pandas as pd
import time

df = pd.read_csv('labeled_and_scored_comments.csv')

In [2]:
df.sort_values(['score'])

id                                       comment_text  \
1158   61a85266de70cca0                         Можешь говорить по русски.   
5246   030a9ba0b32b4456  Rodzina Krusenstern Potomkowie Admirała Adama ...   
11040  a6814193ac6b4705  The opening paragraph states that the game is ...   
24192  f0d6dee96e057a5b  The book is used as a source in a number of ar...   
2332   cbb9268d3c7429c1     Vals Vienés \n |Rumba \n |Cha Cha Chá(ciclo 7)   
...                 ...                                                ...   
19994  1429a2c930688317                          you fucking piece of shit   
28728  d82e45eb103ad07b                   fuck off you stupid aspy asshole   
25535  e9b92c6c6b01aeef  FUCK YOU YOU PIECE OF SHIT \n FUCK YOU YOU PIE...   
25945  e783fd267f3a9d3b  FUCK WIKIPEDIA ON WHEELS! \n\nFuck off wikiped...   
17833  364b6c24210e2345  Fuck you you fucking pig!\n\nYou motherfucking...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  
1158   0.000004      0             0        0       0       0              0  
5246   0.000009      0             0        0       0       0              0  
11040  0.000091      0             0        0       0       0              0  
24192  0.000116      0             0        0       0       0              0  
2332   0.000131      0             0        0       0       0              0  
...         ...    ...           ...      ...     ...     ...            ...  
19994  0.997278      1             1        1       0       1              0  
28728  0.997736      1             1        1       0       1              0  
25535  0.997982      1             1        1       0       1              0  
25945  0.998136      1             1        1       0       1              0  
17833  0.998329      1             0        1       0       1              0  

[41338 rows x 9 columns]

In [8]:
from googleapiclient.discovery import build
import json

def get_toxicity_score(comment):
    
  API_KEY = 'AIzaSyATGcz013tcfM2pcJvuBQBMBudn4G08cFw' 
    
  client = build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  )

  analyze_request = {
  'comment': { 'text': comment },
  'requestedAttributes': {'TOXICITY': {}}
  }
    
  response = client.comments().analyze(body=analyze_request).execute()
  toxicity_score = response["attributeScores"]["TOXICITY"]["summaryScore"]["value"]
    
  return toxicity_score

The thresholds I set were if the score was higher than 0.5, the comment is considered toxic. If the score was higher than 0.75, the comment is considered severely toxic. The reason why I chose these thresholds is because while discovering the excel sheet, I found out that the comments that got a severe_toxic label were likely to score greater than 0.75. 

In [9]:
get_toxicity_score("I'm sorry")

0.023288755

I decided to test the perspective model for bias by putting the comments that I think that recieved a wrong score in the actual code. For example, the comment “I’m sorry” had a score of 0.7910531 on the excel spreadsheet. However, when I put the comment “I’m sorry” in the code in Jupyter, it only received a score of 0.023288755. 

Hypothesis: comments that contain the word "white" has a higher score of toxicity. 
(Following five comments are some example comments that have high scores which prove that my hypothesis is true.)

In [27]:
get_toxicity_score("She is a upper class white racist. She hides it by attacking Black Conservatives.")

0.74195385

In [28]:
get_toxicity_score("I am sure that he must be related to the African Tubins.  If he is a white man, then ALL African-Americans must be whites as well.")

0.36439645

In [29]:
get_toxicity_score(":::::A:s I mentioned, your facts mean nothing in THIS case. But yeah, keep holding on to your beliefs that black men are rapists who just have to rape white women because they can't help themselves (loud snickering) and that white men are the gods of this earth, and that black women are nothing but liars. Whatever makes you feel better.")

0.6769488

This comment represents the case of TP since it supports the hypothesis and the reason why it received a high toxicity score is because of its racist aspect.

In [30]:
get_toxicity_score("YOU BRAINLESS EDITORS You idots are probably majority white. I put a relevant post up called NIGGER CRACKERS, which is a readily used and historically used Racial Epithet, in Northern America and you try to discredit me and pose to ban me from editing.  If the term NIGGER CRACKER can't be posted, then the term NIGGER should be taken down.  If NIGGER isn't taken down, or NIGGER CRACKER isnt allowed up, wikipedia will be facing the biggest law suit of our day and age. Don't threaten my posts again as well. I'm awaiting rapid feedback or my lawyer will be in touch.")

0.8795294

In [35]:
get_toxicity_score("A white fatass that ain't sh!t and acting like a bihcpadding")

0.8393344

This comment represents the case of FP since it does not contain any racist content, but it has a high toxicity score. This comment was marked with a toxic label because of its offensiveness, not racism. 

In [34]:
comment_list = ['hello', 'how are you', 'fine thanks']

for comment in comment_list:
    score = get_toxicity_score(comment)
    print(comment, score)
    time.sleep(1)

hello 0.02397547
how are you 0.018187003
fine thanks 0.014022041


To sum up, these are the processes of finding whether my hypothesis is true or false. I read over the significant amount of comments that contain the word “white” and have high toxicity scores and determined if they have racism in the comment. I also put those comments in the Jupyter code to see if its actual score was high as well. I employed five example comments that have high scores to prove that my hypothesis is true. Those five comments are selected based on its high toxicity scores, and I found out that most of the comments with the word “white” in it contained racism through exploring the dataset. (Of course I explored a lot more than five, but I believe that these five are the best examples to support my hypothesis.)
